rye/src/cli/publish.rs

```diff
@@ -181,30 +194,161 @@ pub fn execute(cmd: Args) -> Result<(), Error> {
     Ok(())
 }
 
-fn prompt_for_token() -> Result<String, Error> {
+struct Credentials {
+    username: Option<String>,
+    password: Option<Secret<String>>,
+}
+
+struct Repository {
+    name: Option<String>,
+    url: Option<Url>,
+}
+
+fn is_unknown_repository(repository: &Repository) -> bool {
+    repository.name.is_none() && repository.url.is_none()
+}
+
+fn is_default_repository(repository: &Repository) -> bool {
+    repository
+        .name
+        .as_ref()
+        .map_or(false, |it| it == DEFAULT_REPOSITORY)
+        || repository
+            .url
+            .as_ref()
+            .map_or(false, |it| it.domain() == Some(DEFAULT_REPOSITORY_DOMAIN))
+}
+
+fn is_keyring_ready(credentials: &Credentials, repository: &Repository) -> bool {
+    credentials.username.is_some() && repository.url.is_some()
```

Comment:
Might also want this as "some username + some url but no password." I'll give it some thought.

Reply:
I'm going to leave keyring independent of password/token.